[ENH] Refactor Catch22 transformer to simplify debugging and use short names in results

julnow opened the pull request:
<!--
Welcome to sktime, and thanks for contributing!
Please have a look at our contribution guide:
https://www.sktime.net/en/latest/get_involved/contributing.html
-->

#### Reference Issues/PRs
<!--
Repulled from https://github.com/sktime/sktime/pull/5989 after syncing forked .branch

Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests.
If no issue exists, you can open one here: https://github.com/sktime/sktime/issues
-->


#### What does this implement/fix? Explain your changes.
<!--
A clear and concise description of what you have implemented.
-->

#### Does your contribution introduce a new dependency? If yes, which one?

<!--
Only relevant if you changed pyproject.toml.
We try to minimize dependencies in the core dependency set. There
are also further specific instructions to follow for soft dependencies.
See here for handling dependencies in sktime: https://www.sktime.net/en/latest/developer_guide/dependencies.html
-->

#### What should a reviewer concentrate their feedback on?

<!-- This section is particularly useful if you have a pull request that is still in development. You can guide the reviews to focus on the parts that are ready for their comments. We suggest using bullets (indicated by * or -) and filled checkboxes [x] here -->

#### Did you add any tests for the change?

<!-- This section is useful if you have added a test in addition to the existing ones. This will ensure that further changes to these files won't introduce the same kind of bug. It is considered good practice to add tests with newly added code to enforce the fact that the code actually works. This will reduce the chance of introducing logical bugs.
-->

#### Any other comments?
<!--
We value all user contributions, no matter how small or complex they are. If you have any questions, feel free to post
in the dev-chat channel on the sktime discord https://discord.com/invite/54ACzaFsn7. If we are slow to review (>3 working days), likewise feel free to ping us on discord. Thank you for your understanding during the review process.
-->

#### PR checklist
<!--
Please go through the checklist below. Please feel free to remove points if they are not applicable.
-->

##### For all contributions
- [ ] I've added myself to the [list of contributors](https://github.com/sktime/sktime/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
  How to: add yourself to the [all-contributors file](https://github.com/sktime/sktime/blob/main/.all-contributorsrc) in the `sktime` root directory (not the `CONTRIBUTORS.md`). Common badges: `code` - fixing a bug, or adding code logic. `doc` - writing or improving documentation or docstrings. `bug` - reporting or diagnosing a bug (get this plus `code` if you also fixed the bug in the PR).`maintenance` - CI, test framework, release.
  See here for [full badge reference](https://allcontributors.org/docs/en/emoji-key)
- [ ] Optionally, I've added myself and possibly others to the [CODEOWNERS](https://github.com/sktime/sktime/blob/main/CODEOWNERS) file - do this if you want to become the owner or maintainer of an estimator you added.
  See here for further details on the [algorithm maintainer role](https://www.sktime.net/en/latest/get_involved/governance.html#algorithm-maintainers).
- [ ] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

##### For new estimators
- [ ] I've added the estimator to the API reference - in `docs/source/api_reference/taskname.rst`, follow the pattern.
- [ ] I've added one or more illustrative usage examples to the docstring, in a pydocstyle compliant `Examples` section.
- [ ] If the estimator relies on a soft dependency, I've set the `python_dependencies` tag and ensured
  dependency isolation, see the [estimator dependencies guide](https://www.sktime.net/en/latest/developer_guide/dependencies.html#adding-a-soft-dependency).

<!--
Thanks for contributing!
-->

julnow commented:
@yarnabrina @fkiraly I had to reopen it after syncing, as the last one closed automatically.

It's still work in progress, I have to implement proper naming of the columns (which was the main goal). 

I also think of passing a dict of variables to `_transform_case` methods, and creating a wrapper for the original functions which would unpack needed variables from **kwargs passed to each `_get_feature_function(feature)(*args)` call. What do you think about it @fkiraly ?

julnow commented:
Also, what about https://github.com/sktime/sktime/issues/5990?

fkiraly commented:
> Also, what about https://github.com/sktime/sktime/issues/5990

This also applies to `Catch22` - it implements an unnecessary loop across time series instances - and could be much simplified.

I've done the same recently with `Catch22Wrapper`: https://github.com/sktime/sktime/pull/5983, the recipe is:

* change tags: `X_inner_mtype` to `pd.Series` or `pd.DataFrame` (whichever more convenient), change `univariate-only` to True
* remove all loops over instances and variables
* deprecate `n_jobs`, add the deprecation message, switch use of `n_jobs` to use `set_config` for the time of the deprecation

It's up to you whether you do this in one PR, separate PR, or not at all - it's probably easier to separate this out, if you're changing logic as well.

fkiraly commented:
FYI, linting (code formatting) checks are failing, see here for a guide how to ensure you follow these, and how to set up the checks locally: https://www.sktime.net/en/stable/developer_guide/coding_standards.html

julnow commented:
> FYI, linting (code formatting) checks are failing, see here for a guide how to ensure you follow these, and how to set up the checks locally: https://www.sktime.net/en/stable/developer_guide/coding_standards.html

Yes, I change the typing from `list` to `List` so I hope it works now.




> > Also, what about #5990
> 
> This also applies to `Catch22` - it implements an unnecessary loop across time series instances - and could be much simplified.
> 
> I've done the same recently with `Catch22Wrapper`: #5983, the recipe is:
> 
> * change tags: `X_inner_mtype` to `pd.Series` or `pd.DataFrame` (whichever more convenient), change `univariate-only` to True
> * remove all loops over instances and variables
> * deprecate `n_jobs`, add the deprecation message, switch use of `n_jobs` to use `set_config` for the time of the deprecation
> 
> It's up to you whether you do this in one PR, separate PR, or not at all - it's probably easier to separate this out, if you're changing logic as well.

My only concern is that right now I use it in my project for multivariate time series, with a   `nested_univ` dataframe as input. How would you recommend dealing it with it? Should the implementation allow the multivariate dataframe, or it would be better to run this in loop and concat columns in the end? Not very elegant I'm afraid.

Output like in https://www.sktime.net/en/stable/examples/transformation/feature_extraction_with_tsfresh.html (Multivariate time series classification data) would be ideal for me. - InputColumnName_FeatureName (or its index like in your implementation here: https://github.com/sktime/sktime/pull/5983)

fkiraly commented:
> My only concern is that right now I use it in my project for multivariate time series, with a nested_univ dataframe as input.

Can you explain how/why this would pose a problem? You can still pass nested dataframes or other `sktime` compatible containers after the proposed refactor, it's just that the iteration would be handled by the base class boilerplate.

fkiraly commented:
(FYI, linting again - it says it wants double quotes instead of single quotes for strings)

julnow commented:
> > My only concern is that right now I use it in my project for multivariate time series, with a nested_univ dataframe as input.
> 
> Can you explain how/why this would pose a problem? You can still pass nested dataframes or other `sktime` compatible containers after the proposed refactor, it's just that the iteration would be handled by the base class boilerplate.
> 
> The column naming is also automatic, it is `variablename__featurename` if the iteration is done by the base class.

Okay great I see it now, thanks!

julnow commented:
@fkiraly I added a refactor of `_catch22_numba` to pass values as a numba-type Dict, but I'm stuck on `module numba not found` error, do you know how to fix it?

fkiraly commented:
@julnow, can you describe when you get the `numba` related error?

You should ensure that tests run here on the remote PR, for that you will need to pass the code formatting checks in `code-quality`. See link to guide here: https://github.com/sktime/sktime/pull/6002#issuecomment-1962999340

fkiraly commented on sktime/transformations/panel/catch22wrapper.py:
```diff
@@ -126,8 +126,6 @@ def __init__(
         else:
             raise ValueError("features must be a str, list or tuple")
 
-        self._transform_features = None
```
this is concerning the other estimator - can you explain why you are changing this?
You are probably right, this can be removed.

Even if, it would be better if you could split this out in a separate pull request? The changes seem independent enough, so that's better for testing, review by core developers, and for the release changelog.

julnow replied:
I didn't see where it is used, so I wanted to remove the additional `if not transform_feature[f_count]: continue` if condition when I was changing the implementation, but I can bring it back of course

fkiraly replied:
you're probably right in removing it, I wondered what it was doing, closer inspection reveals, probably nothing. Still, it's easy to miss in a PR focusing on sth else, that's why I suggest simply moving this to another PR.

julnow replied:
sure will do

julnow replied:
Brought back in https://github.com/sktime/sktime/pull/6002/commits/1b04ebcacd6dc4f386f6fd4d075d9ad9342f3929

fkiraly commented on sktime/utils/validation/__init__.py:
```diff
@@ -142,6 +142,8 @@ def check_n_jobs(n_jobs: int) -> int:
         return 1
     elif not is_int(n_jobs):
         raise ValueError(f"`n_jobs` must be None or an integer, but found: {n_jobs}")
+    elif n_jobs == -1:
```
FYI @yarnabrina, due to recent discussion on `check_n_jobs`. Do you have a link to the issue I am trying to find? Search for `check_n_jobs` in the issue tracker does not find anything.

julnow replied:
@fkiraly @yarnabrina it brings another error:
```
AttributeError  
      1 cif = CanonicalIntervalForest(n_estimators=50, att_subsample_size=8, random_state=47)
----> 2 cif.fit(X_train,y_train)
      4 cif_preds = cif.predict(X_test)
      5 print("CIF Accuracy: " + str(metrics.accuracy_score(y_test, cif_preds)))

File /opt/hostedtoolcache/Python/3.9.18/x64/lib/python3.9/site-packages/sktime/classification/base.py:240, in BaseClassifier.fit(self, X, y)
    235         raise AttributeError(
    236             "self.n_jobs must be set if capability:multithreading is True"
    237         )
    239 # pass coerced and checked data to inner _fit
--> 240 self._fit(X,y)
    241 self.fit_time_ = int(round(time.time() * 1000)) - start
    243 # this should happen last: fitted state is set to True
```

julnow replied:
However, it is due to implementation of `Catch22Classifier`, so I don't know if it should be managed in this PR?

fkiraly replied:
In general, no dependent code should be broken by any pull request, without sufficient forewarning (see depercation policy: https://www.sktime.net/en/stable/developer_guide/deprecation.html)

I think there's three options:

* we make sure `Catch22` behaves identically for all input combinations pre and post this PR
* the reworked estimator is added as a third catch-22 estimator
* the reworked estimator is added as a third catch-22 estimator, and we run a single deprecation/change cycle to switch names in some sensible way

My preference would be option 1 if possible, since the PR claims to be a refactor plus extension of functionality, so if this is the case, current calls should not be impacted.

fkiraly replied:
if the change in `check_n_jobs` would change/break dependent code, we should not make it in this PR, but dedicate change/deprecation ot it. If @yarnabrina can help with finding the other discussion we had on this, the gist of it is that `check_n_jobs` occurs only in places where imo it is not necessary, and it could be refactored away, in favour of using `utils.parallel`.

yarnabrina replied:
Maybe you are looking for #5990 ?

fkiraly replied:
yes, exactly.

julnow commented:
> @julnow, can you describe when you get the `numba` related error?
> 
> You should ensure that tests run here on the remote PR, for that you will need to pass the code formatting checks in `code-quality`. See link to guide here: [#6002 (comment)](https://github.com/sktime/sktime/pull/6002#issuecomment-1962999340)

code-quality is passing now, but I cannot force it to import 
`from numba import types
    from numba.typed import Dict`

you can see that it is failing in run-notebook-examples, and I'm stuck as I cannot figure out how to import this. In under modules I  saw that `from numba.typed import Dict` is imported inside a method, but it's not helping in my example.

I could revert to this conditionional passing of right arguments to numba methods, but I thought this would be more readable if passed as `numba.typed.Dict`

julnow commented:
@fkiraly 
for
`@njit
def _create_numba_dict(
    series: np.ndarray,
    smin: float,
    smax: float,
    smean: float,
    std: float,
    outlier_series: np.ndarray,
    ac: np.ndarray,
    acfz: int,
):
    from numba import types
    from numba.typed import Dict

    numba_dict = Dict()
    numba_dict["series"] = series
    numba_dict["smin"] = np.array([smin], dtype=types.float64)
    numba_dict["smax"] = np.array([smax], dtype=types.float64)
    numba_dict["smean"] = np.array([smean], dtype=types.float64)
    numba_dict["std"] = np.array([std], dtype=types.float64)
    numba_dict["outlier_series"] = outlier_series
    numba_dict["ac"] = ac
    numba_dict["acfz"] = np.array([acfz], dtype=types.float64)
    return numba_dict`

I get `File /opt/hostedtoolcache/Python/3.9.18/x64/lib/python3.9/site-packages/numba/core/dispatcher.py:471, in _DispatcherBase._compile_for_args(self, *args, **kws)
    468     error_rewrite(e, 'typing')
    469 except errors.UnsupportedError as e:
    470     # Something unsupported is present in the user code, add help info
--> 471     error_rewrite(e,'unsupported_error')
    472 except (errors.NotDefinedError, errors.RedefinedError,
    473         errors.VerificationError) as e:
    474     # These errors are probably from an issue with either the code
    475     # supplied being syntactically or otherwise invalid
    476     error_rewrite(e, 'interpreter')

File /opt/hostedtoolcache/Python/3.9.18/x64/lib/python3.9/site-packages/numba/core/dispatcher.py:409, in _DispatcherBase._compile_for_args.<locals>.error_rewrite(e, issue_type)
    407     raise e
    408 else:
--> 409     raise e.with_traceback(None)

UnsupportedError: Failed in nopython mode pipeline (step: analyzing bytecode)
Use of unsupported opcode (IMPORT_NAME) found

File "../../../../../../opt/hostedtoolcache/Python/3.9.18/x64/lib/python3.9/site-packages/sktime/transformations/panel/_catch22_numba.py", line 949:
def _create_numba_dict(
    <source elided>
):
    from numba import types
    ^


Error: Process completed with exit code 1.`

fkiraly commented:
I see, it looks like `numba` is unhappy if you import inside an `njit` compliled function.

The key problem is to ensure that `sktime` works without `numba`, so there can be no imports of `numba` in locations accessed throughout, e.g., crawling. This is also sometimes referred to as "soft dependency isolation", as it makes sure that `numba` is not necessary as a dependency for `sktime` to "not break".

A way that could satisfy both conditions (a) no import inside `njit` compiled fun, and (b) no import of `numba` if not present is via runtime dependency check.

Examples:

* `classification.dictionary_based._tde_numba`, line 8
* `classification.interval_based._rise_numba`, line 10
* `transformations.panel.dictionary_based._sfa_fast_numba`, line 14, this has the same context of `typed.Dict`

PS, for multiline code blocks, use three backticks/graves, not one (fixed above)

fkiraly commented on examples/catch22.ipynb:
```diff
@@ -30,7 +30,7 @@
   },
   {
    "cell_type": "code",
-   "execution_count": null,
```
updating hte notebook is fine, just make sure you have reset and run it only once, in the final commit.

julnow replied:
got it, I rerun and committed now

yarnabrina commented:
I don't think I should comment as I have zero familiarity with numba, but if you add another level what happens?

What I mean is you make the import in a method, and then define an internal function and call it? Will that support compilation restrictions? 

(Strong emphasis on zero experience on numba part, apologies if this suggestion is not useful)

fkiraly commented:
> What I mean is you make the import in a method, and then define an internal function and call it? Will that support compilation restrictions?

I would suspect that does not work, `numba` compilation seems to require "flat" code, but that's only from experience in seeing working code rather than writing a lot of it, or a genuine understanding of what's going on under the hood.

fkiraly reviewed (changes requested):
Would you be so kind and move the changes to `Catch22Wrapper` to a separate pull request?

You can do so by branching off `main` in your repo, and PR from that branch.

Happy to help with managing multiple branches if you haven't done that before.

julnow commented:
> Would you be so kind and move the changes to `Catch22Wrapper` to a separate pull request?
> 
> You can do so by branching off `main` in your repo, and PR from that branch.
> 
> Happy to help with managing multiple branches if you haven't done that before.
@fkiraly 
I will, but won't it cause tests to fail as  `Catch22Wrapper` depends on feature names defined in `Catch22` which are changed here?

Also, I can run the notebooks locally but they keep failing in ci/cd, would you have a moment to look at that?

julnow commented:
> > Would you be so kind and move the changes to `Catch22Wrapper` to a separate pull request?
> > You can do so by branching off `main` in your repo, and PR from that branch.
> > Happy to help with managing multiple branches if you haven't done that before.
> > @fkiraly
> > I will, but won't it cause tests to fail as  `Catch22Wrapper` depends on feature names defined in `Catch22` which are changed here?
> 
> Also, I can run the notebooks locally but they keep failing in ci/cd, would you have a moment to look at that?

Once it all works and i have final version of catch22, I will branch off my main (if it's okay)).

fkiraly commented:
> I will, but won't it cause tests to fail as  `Catch22Wrapper` depends on feature names defined in `Catch22` which are changed here?

Hm, I see - if you are changing the feature names, you should copy them over.
We sould also make sure no current code change behaviour in `Catch22`.

> Also, I can run the notebooks locally but they keep failing in ci/cd, would you have a moment to look at that?

Sure! The first stage would be to make sure that `check_estimator` is passing, for `Catch22` and the dependent `Catch22Classifier`.

fkiraly commented:
oops, it looks like an accident happened, and a lot of files were now committed to this PR.

Could you kindly check, @julnow, and remove unnecessary file from the PR?

fkiraly commented:
yes, now it looks like it's fixed

julnow commented:
And tests are passing locally now, I hope it will pass here as well

fkiraly commented:
Yes, that's related to this: https://github.com/sktime/sktime/issues/6040
(exception type changed by accident - should not impact users, but will fix quickly)

julnow commented:
> Would you be so kind and move the changes to `Catch22Wrapper` to a separate pull request?
> 
> You can do so by branching off `main` in your repo, and PR from that branch.
> 
> Happy to help with managing multiple branches if you haven't done that before.

Okay, can this be closed? Given:

> I also removed excess changes from `Catch22Wrapper` leaving only change from `catch22.feature_names` to `catch22.FEATURE_NAMES`

fkiraly commented:
Yes, I'll restart the tests and review properly in a moment.

fkiraly commented on sktime/transformations/panel/_catch22_numba.py:
```diff
@@ -487,26 +487,38 @@ def _spline_fit(X):
+
+
+@njit(fastmath=True, cache=True)
+def _DN_HistogramMode_5(series, smin, smax, smean, std, outlier_series, ac, acfz):
```
can you explain why you give all features now eight arguments, `series`, `smin`, etc? Is this for performance reasons?

It is less crisp design-wise, as seven of these depend on the first.

If for performance reasons, have you tested performance?

julnow replied:
I wanted to omit passing args based on feature_id as in original `_transform_case` method, as it seemed overcomplicated and hard to debug. I tested using `numba.typed.Dict` but it was failing for multiple reasons. In theory only series and outlied_series could be passed, and the rest could be calculated later, but e.g., `_autocorellation` cannot be run in numba non-python mode, so then a wrapper for a method using this parameter should be created etc. So I thought it's better to pass all arguments for the time being, and improve it in the future. It's still better than:
```
if feature == 0 or feature == 1 or feature == 11:
                    if smin is None:
                        smin = np.min(series)
                    if smax is None:
                        smax = np.max(series)
                    args = [series, smin, smax]
                elif feature == 2 or feature == 22:
                    if smean is None:
                        smean = np.mean(series)
                    args = [series, smean]
                elif feature == 3 or feature == 4:
                    if self.outlier_norm:
                        if smean is None:
                            smean = np.mean(series)
                        if outlier_series is None:
                            outlier_series = _normalise_series(series, smean)
                        args = [outlier_series]
                    else:
                        args = [series]
                elif feature == 7 or feature == 8:
                    if smean is None:
                        smean = np.mean(series)
                    if fft is None:
                        nfft = int(
                            np.power(2, np.ceil(np.log(len(series)) / np.log(2)))
                        )
                        fft = np.fft.fft(series - smean, n=nfft)
                    args = [series, fft]
                elif feature == 5 or feature == 6 or feature == 12:
                    if smean is None:
                        smean = np.mean(series)
                    if fft is None:
                        nfft = int(
                            np.power(2, np.ceil(np.log(len(series)) / np.log(2)))
                        )
                        fft = np.fft.fft(series - smean, n=nfft)
                    if ac is None:
                        ac = _autocorr(series, fft)
                    args = [ac]
                elif feature == 16 or feature == 17 or feature == 20:
                    if smean is None:
                        smean = np.mean(series)
                    if fft is None:
                        nfft = int(
                            np.power(2, np.ceil(np.log(len(series)) / np.log(2)))
                        )
                        fft = np.fft.fft(series - smean, n=nfft)
                    if ac is None:
                        ac = _autocorr(series, fft)
                    if acfz is None:
                        acfz = _ac_first_zero(ac)
                    args = [series, acfz]
```
in my opinion

fkiraly replied:
I see. Yes, your first try was my inuition (see `Catch22Wapper`), but I see it interacts badly with `numba`. Fine with me for now, then.

Perhaps adding a short few lines of comments at the top of the block, about function signature, would be helpful to general developers looking at the file.

fkiraly replied:
> It's still better than:

Yes, of course.

julnow replied:
> I see. Yes, your first try aligns with my design intuition (see `Catch22Wapper`), but I see it interacts badly with `numba`. Fine with me for now, then.
> 
> Perhaps adding a short few lines of comments at the top of the block, about function signature, would be helpful to general developers looking at the file.

Sure, I'll add it! Passing it though a dict would be better I think, but e.g., every type of dict (in numba) must be the same, so e.g., a float has to be saved as single-element list of floats if other elements are floats. I'd have to delve into it

fkiraly reviewed (changes requested):
Nice!

Requests/ questions:

* if you already introduce short feature names, should it not be possible for users to pass the short feature names as well, to `features`? That is probably not much work and can be achieved by inverting `SHORT_FEATURE_NAMES_DICT`
* a question above about the changed signature of the feature functions - not necessarily sth to change, though I would like to understand.
* `get_test_params` should properly cover the range of parameters we can select. Currently this is a bit underdeveloped (so, dangerous due to low coverage)

Non-blocking:

* the docstring does not render properly, esp the list of featue names.
* should therer be an `"auto"` option too for `col_names`? To return exactly `features` as column names in the same order.
* if we allow for this, mixed type `features` (int, short, long) should be tested

julnow commented:
> Nice!
> 
> Requests/ questions:
> 
> * if you already introduce short feature names, should it not be possible for users to pass the short feature names as well, to `features`? That is probably not much work and can be achieved by inverting `SHORT_FEATURE_NAMES_DICT`

Yes, I think I can add it to this PR easily
> * a question above about the changed signature of the feature functions - not necessarily sth to change, though I would like to understand.

`function` seemed less vague than `fun`, but I can change it to `fun` if it's a convention
> * `get_test_params` should properly cover the range of parameters we can select. Currently this is a bit underdeveloped (so, dangerous due to low coverage)

Sure, I'll add it as well
> 
> Non-blocking:
> 
> * the docstring does not render properly, esp the list of featue names.

Okay I'll correct it, I'm more used to google than numpy convention, but I'll figure it out :)
> * should there be an `"auto"` option too for `col_names`? To return exactly `features` as column names in the same order.

Okay, I think it could be added, maybe in the scope of next PR?
> * if we allow for this, mixed type `features` (int, short, long) should be tested

fkiraly commented:
> `function` seemed less vague than `fun`, but I can change it to `fun` if it's a convention

Did't mean the name, signature also includes args. Basically what we discussed above, I consider that point resolved:
https://github.com/sktime/sktime/pull/6002#discussion_r1508865511

fkiraly commented:
> Okay I'll correct it, I'm more used to google than numpy convention, but I'll figure it out :)

You can check here how it renders:
![image](https://github.com/sktime/sktime/assets/7985502/37495149-c672-4341-a441-3253e98a1be1)

fkiraly commented:
>> should there be an "auto" option too for col_names? To return exactly features as column names in the same order.

> Okay, I think it could be added, maybe in the scope of next PR?

Sure - I thought it would be as easy as doing a `return self.features` in `_prepare_output_col_names`, so I'm wondering where you see the difficulty (I might be overlooking sth)

julnow commented:
> > > should there be an "auto" option too for col_names? To return exactly features as column names in the same order.
> 
> > Okay, I think it could be added, maybe in the scope of next PR?
> 
> Sure - I thought it would be as easy as doing a `return self.features` in `_prepare_output_col_names`, so I'm wondering where you see the difficulty (I might be overlooking sth)

Oh you're absolutely right, I'll add it here.

fkiraly commented:
it's also fine as-is - let us know when ready for review.

julnow commented:
> it's also fine as-is - let us know when ready for review.

I think it is already @fkiraly @yarnabrina

fkiraly reviewed (changes requested):
Excellent! Some comments.

Blocking (for discussion or change):

* `check_n_jobs` is no longer used, and the change seems to affect logic. I would say, let's remove it from this PR, and if you think it is needed, discuss in another PR or issue.


Non-blocking:

* why do we need two dictionaries, `METHODS_DICT` and `CATCH24_METHODS_DICT` - would it not simplify multiple places? Similarly, for short feature names. Odd that these are/were treated separately.
* docstring should mention this also supports catch-24 features (plus mean, std)
* why is the `catch24` argument needed and mean/std treated completely differently? That is confusing. Why not simply add mean and std as additional features to the large list? Needs to be deprecate, of course.
* `outlier_norm` docstring should mention which two feaures it affects.
* the `_transform_single_feature` method is very unpleasant, as it is private, yet used in random classifiers. Not necessary for this PR, but a question: can you come up with a good way to decouple this, e.g., joint utility dispatcher or similar? It is not good style if estimtors keep calling each others' private methods, and risky. As a warning, if it is not immediately obvious, I would recommend not looking into it as things like this can develop into time consuming rabbit holes...

julnow commented:
> Excellent! Some comments.
> 
> Blocking (for discussion or change):
> 
> * `check_n_jobs` is no longer used, and the change seems to affect logic. I would say, let's remove it from this PR, and if you think it is needed, discuss in another PR or issue.

Okay, removed
> 
> Non-blocking:
> 
> * why do we need two dictionaries, `METHODS_DICT` and `CATCH24_METHODS_DICT` - would it not simplify multiple places? Similarly, for short feature names. Odd that these are/were treated separately.

I think it's due to the fact that std and mean weren't mentioned in the original publication: https://link.springer.com/article/10.1007/s10618-019-00647-x and were added later, so are not really a core part of catch22. That's why I created two seperate dicts, but this could be discussed in the future, but all the other packages (pycatch22, julia one etc.) also seperate the logic of core 22 features and these additional two.

> * docstring should mention this also supports catch-24 features (plus mean, std)

added

> * why is the `catch24` argument needed and mean/std treated completely differently? That is confusing. Why not simply add mean and std as additional features to the large list? Needs to be deprecate, of course.

Same as above, maybe features should be "catch22"  by default or "catch24" instead of "all", but this could be changed later in my opinion.


> * `outlier_norm` docstring should mention which two feaures it affects.

Added
> * the `_transform_single_feature` method is very unpleasant, as it is private, yet used in random classifiers. Not necessary for this PR, but a question: can you come up with a good way to decouple this, e.g., joint utility dispatcher or similar? It is not good style if estimtors keep calling each others' private methods, and risky. As a warning, if it is not immediately obvious, I would recommend not looking into it as things like this can develop into time consuming rabbit holes...

I spent a long time with this method which is in fact not necessary in the `catch22` transformer, but I left it for compatibility with the other module. This should definetely be resolved in the future, but I didn't have time - the other module should be accordingly rewritten.

fkiraly commented:
> so are not really a core part of catch22. That's why I created two seperate dicts, but this could be discussed in the future, but all the other packages (pycatch22, julia one etc.) also seperate the logic of core 22 features and these additional two.

I am aware, there's catch22 and catch24. What I mean, on the code side there doesn't need to be a distinction, they are just 24 features and sometimes people only want 22 of those. Imo there's no reason to internally separate dictionaries etc, there still can be a separation in the user interface, or having only 22 as the default.

fkiraly commented:
> Same as above, maybe features should be "catch22" by default or "catch24" instead of "all", but this could be changed later in my opinion.

Sure - when I say "non-blocking", it means, fine if changed later or not at all.

fkiraly commented:
> I spent a long time with this method which is in fact not necessary in the catch22 transformer ... This should definetely be resolved in the future, but I didn't have time

Agreed - also non-blocking, as said. Just in case you had an idea on how to resolve this quickly, reading this I see that it probably is not that easy to clean this up.

julnow commented:
> I am aware, there's catch22 and catch24. What I mean, on the code side there doesn't need to be a distinction, they are just 24 features and sometimes people only want 22 of those. Imo there's no reason to internally separate dictionaries etc, there still can be a separation in the user interface, or having only 22 as the default.

Okay, do you think it could be done in the next PL in parallel with modifying the selection of features (removing `catch24` bool argument)?

fkiraly commented:
> Okay, do you think it could be done in the next PR

Sure - when it's listed "non-blocking", it means it does not prevent from getting this PR merged if not addresesd, and can well be addressed in another PR, or not a tall.

fkiraly commented on sktime/transformations/panel/catch22.py:
```diff
@@ -603,4 +498,7 @@ def get_test_params(cls, parameter_set="default"):
-        return [param1, param2]
+        param3 = {"features": [1, 12, 21]}
+        param4 = {"features": ["forecast_error", "centroid_freq"]}
+        param5 = {"features": [11, "DN_HistogramMode_5", "forecast_error"]}
```
here we should cover the other parameters too. There are a couple. `n_jobs` is not needed, ofc.

I think we can get good coverage with 5 sets, just add one version each, so each value is used somewhere, no need to multiply exponentially.

fkiraly reviewed (changes requested):
Looks great to me!

Two small but blocking requests:

* parameter sequence, see above
* parameter coverage in tests, see above

fkiraly reviewed (approved):
Thanks - all addressed.

Once (or - if) tests pass, we can merge, from my side.

julnow commented:
@fkiraly I have one last question, as I cannot find it in documentation -- how should the n_jobs be passed in non-deprecated way?

fkiraly commented:
the warning and docstring should tell you - via `set_config` and `backend:params`. Please raise an issue if you think that information is not easily accessible or obscure, because that would be a problem (and a place for improvement).
